# Практичні завдання з pandas для початківців

**Мета:** Набути базових навичок роботи з бібліотекою pandas для аналізу табличних даних.

**Інструкції:**
- Виконуйте завдання послідовно — кожне наступне спирається на попередні.
- Після виконання кожного завдання напишіть короткий висновок у комірці `**Висновок:**`.
- Не змінюйте код у комірках з підготовкою даних.

**Набір даних:** Умовна таблиця продажів інтернет-магазину за 2023 рік.

---
## Підготовка: Генерація тестових даних

Запустіть цю комірку перед початком роботи. Не змінюйте код.

In [1]:
import pandas as pd
import numpy as np

np.random.seed(42)
n = 200

categories = ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт']
cities = ['Київ', 'Харків', 'Одеса', 'Дніпро', 'Львів']
payment_methods = ['Картка', 'Готівка', 'Онлайн']

df = pd.DataFrame({
    'order_id':        range(1001, 1001 + n),
    'date':            pd.date_range('2023-01-01', periods=n, freq='D')[:n],
    'category':        np.random.choice(categories, n),
    'product':         ['Товар_' + str(i) for i in np.random.randint(1, 51, n)],
    'city':            np.random.choice(cities, n),
    'quantity':        np.random.randint(1, 10, n),
    'price':           np.round(np.random.uniform(50, 5000, n), 2),
    'payment_method':  np.random.choice(payment_methods, n),
    'rating':          np.random.choice([1, 2, 3, 4, 5, np.nan], n, p=[0.05, 0.1, 0.15, 0.3, 0.3, 0.1]),
    'is_returned':     np.random.choice([True, False], n, p=[0.1, 0.9]),
})

# Додаємо кілька пропущених значень
df.loc[np.random.choice(df.index, 10, replace=False), 'city'] = np.nan
df.loc[np.random.choice(df.index, 5, replace=False), 'price'] = np.nan

# Розраховуємо загальну суму
df['total'] = df['quantity'] * df['price']

print('Дані успішно згенеровано. Рядків:', len(df))
print('Колонки:', list(df.columns))

Дані успішно згенеровано. Рядків: 200
Колонки: ['order_id', 'date', 'category', 'product', 'city', 'quantity', 'price', 'payment_method', 'rating', 'is_returned', 'total']


---
## Завдання 1 (Легке). Перший погляд на дані

Виведіть перші 10 рядків датафрейму `df` та отримайте загальну інформацію про нього (кількість рядків, колонок, типи даних) за допомогою методів `head()` та `info()`.

In [2]:
# Ваш код тут

print(df.head(10))

df.info()

   order_id       date  category   product    city  quantity    price  \
0      1001 2023-01-01  Продукти  Товар_34    Київ         1   963.38   
1      1002 2023-01-02     Спорт   Товар_6   Львів         6  1086.28   
2      1003 2023-01-03     Книги  Товар_22  Дніпро         4  1883.84   
3      1004 2023-01-04     Спорт  Товар_11  Дніпро         1  2448.39   
4      1005 2023-01-05     Спорт  Товар_48  Дніпро         7  3110.36   
5      1006 2023-01-06      Одяг  Товар_16   Львів         9  1876.12   
6      1007 2023-01-07     Книги  Товар_33  Дніпро         4  2339.55   
7      1008 2023-01-08     Книги   Товар_9   Львів         4  3749.98   
8      1009 2023-01-09     Книги   Товар_6  Дніпро         6   231.58   
9      1010 2023-01-10     Спорт  Товар_16   Одеса         3  1299.56   

  payment_method  rating  is_returned     total  
0         Картка     5.0        False    963.38  
1         Картка     4.0        False   6517.68  
2        Готівка     5.0        False   7535.3

**Висновок:** *(Напишіть, скільки рядків і колонок у датафреймі, які типи даних зустрічаються, чи є пропущені значення)*

Датафрейм містить 200 рядків та 11 колонок. У ньому присутні такі типи даних:  

`int64`,   
`float64`,   
`object`,   
`datetime64[ns]`,   
`bool`.  

У датафреймі є пропущені значення:  

у колонці `city` — 10 пропусків;  
у колонці `price` — 5 пропусків;  
у колонці `rating` — 23 пропусків;  
у колонці  `total` — 5 пропусків.  

Тому перед подальшим аналізом даних варто перевірити та за потреби обробити пропущені значення.  

---
## Завдання 2 (Легке). Описова статистика

Використайте метод `describe()` для отримання описової статистики числових колонок. Окремо виведіть мінімум, максимум і середнє значення колонки `price`.

In [3]:
# Ваш код тут

print("Мінімум price:", df['price'].min())
print("Максимум price:", df['price'].max())
print("Середнє price:", df['price'].mean())

print("Медіана price:", df['price'].median())

df.describe()

Мінімум price: 139.65
Максимум price: 4950.3
Середнє price: 2500.579641025641
Медіана price: 2517.29


,order_id,date,quantity,price,rating,total
count,200.000000,200,200.000000,195.000000,177.000000,195.000000
mean,1100.500000,2023-04-10 12:00:00,4.700000,2500.579641,3.813559,11629.097590
min,1001.000000,2023-01-01 00:00:00,1.000000,139.650000,1.000000,149.350000
25%,1050.750000,2023-02-19 18:00:00,2.000000,1288.215000,3.000000,3893.955000
50%,1100.500000,2023-04-10 12:00:00,5.000000,2517.290000,4.000000,8746.840000
75%,1150.250000,2023-05-30 06:00:00,7.000000,3706.975000,5.000000,16922.420000
max,1200.000000,2023-07-19 00:00:00,9.000000,4950.300000,5.000000,44324.640000
std,57.879185,NaN,2.618066,1380.066748,1.169808,9968.293697


**Висновок:** *(Опишіть діапазон цін, середнє значення, чи є суттєвий розкид)*  
Ціни в даному наборі даних варіюються від 139.65 до 4950.30, що свідчить про широкий діапазон значень. Середня ціна становить приблизно 2500.58, тобто більшість товарів знаходяться в середньому ціновому сегменті.  

Медіана (50%) дорівнює 2517.29, що дуже близько до середнього значення. Це означає, що розподіл цін є відносно симетричним, без сильного перекосу в один бік.  

Стандартне відхилення становить приблизно 1380, що є доволі високим значенням. Це вказує на суттєвий розкид цін — товари сильно відрізняються за вартістю (є як дешеві, так і дорогі позиції).  

**Підсумок:**  
- Мінімальна ціна: 139.65  
- Максимальна ціна: 4950.30  
- Середня ціна: ~2500.58  
- Розкид: значний (висока варіативність цін)  

---
## Завдання 3 (Легке). Пропущені значення

Знайдіть кількість пропущених значень у кожній колонці за допомогою `isnull().sum()`. Визначте, який відсоток рядків має пропущені значення хоча б в одній колонці (`dropna()` та `len()`).

In [4]:
# Ваш код тут
df.isnull().sum()



,0
order_id,0
date,0
category,0
product,0
city,10
quantity,0
price,5
payment_method,0
rating,23
is_returned,0


In [5]:
total_cells = df.size
missing_cells = df.isnull().sum().sum()

print("Всього ячейок:", total_cells)
print("Пропущених ячейок:", missing_cells)
print("Відсоток пропусків:", missing_cells / total_cells * 100)

Всього ячейок: 2200
Пропущених ячейок: 43
Відсоток пропусків: 1.9545454545454546


**Висновок:** *(У яких колонках є пропуски, наскільки це критично для аналізу)*  
Пропущені значення розподілені по окремих колонках і становлять невелику частку від загальної кількості ячейок у датасеті. Загальний відсоток пропусків є низьким, тому дані в цілому придатні для аналізу, але потребують часткової обробки пропущених значень у колонках `city`, `price` та `rating`.

---
## Завдання 4 (Легке). Відбір колонок і рядків

Створіть новий датафрейм `df_subset`, що містить лише колонки `order_id`, `category`, `city`, `total`. Потім виведіть рядки з індексами від 10 до 20 включно за допомогою `.iloc[]`.

In [6]:
# Ваш код тут
df_subset = df[['order_id', 'category', 'city', 'total']]

df_subset.iloc[10:21]

,order_id,category,city,total
10,1011,Продукти,Дніпро,21486.48
11,1012,Книги,Львів,31368.89
12,1013,Спорт,Харків,7748.40
13,1014,Одяг,Дніпро,18787.72
14,1015,Продукти,NaN,1741.50
15,1016,Одяг,Одеса,4529.38
16,1017,Продукти,Київ,10745.84
17,1018,Спорт,Одеса,10001.84
18,1019,Електроніка,Дніпро,12444.75
19,1020,Продукти,Харків,13422.92


**Висновок:** *(Поясніть різницю між вибором колонок через список і через `iloc`)*  
Вибір колонок через список `(df[['order_id', 'category', 'city', 'total']])` дозволяє звертатися до даних за назвами колонок, тобто явно вказувати, які ознаки потрібні. Це більш читабельний і безпечний спосіб, оскільки не залежить від порядку колонок у датафреймі.  

Метод `.iloc[]` використовується для вибору даних за позицією (індексами). Наприклад, `df_subset.iloc[10:21]` означає вибір рядків з 10 по 20 включно за їх порядковим номером, незалежно від назв.  

Отже:  

вибір через список — працює з іменами колонок;  
`.iloc[]` — працює з числовими індексами (позиціями).

---
## Завдання 5 (Легке). Фільтрація даних

Відфільтруйте датафрейм і отримайте:
1. Всі замовлення з категорії `'Електроніка'`.
2. Всі замовлення, де `price` перевищує 2000.
3. Поверненi замовлення (`is_returned == True`) з міста `'Київ'`.

Виведіть кількість рядків для кожного фільтру.

In [7]:
# Ваш код тут
electronics = df[df['category'] == 'Електроніка']
price_gt_2000 = df[(df['category'] == 'Електроніка') & (df['price'] > 2000)]
kyiv_returns = df[
    (df['category'] == 'Електроніка') &
    (df['city'] == 'Київ') &
    (df['is_returned'] == True)
]
kyiv_orders = df[df['city'] == 'Київ']
return_rate = len(kyiv_returns) / len(kyiv_orders) * 100


print("Електроніка:", len(electronics))
print("Електроніка Price > 2000:", len(price_gt_2000))
print("Повернення з Києва:", len(kyiv_returns))
print("Повернення з Києва (%):", return_rate)

Електроніка: 43
Електроніка Price > 2000: 23
Повернення з Києва: 2
Повернення з Києва (%): 5.405405405405405


**Висновок:** *(Яка частка замовлень відповідає кожному критерію? Що можна сказати про повернення з Києва?)*  

Аналіз даних по категорії «Електроніка» показує, що вона є важливою частиною продажів — всього 43 замовлення, з яких 23 належать до дорогого сегмента (price > 2000). Це означає, що більше половини покупок припадає на високовартісні товари, що формує сильний преміум-попит.  

Рівень повернень у Києві становить лише 5.4% (2 повернення), що є низьким показником і свідчить про достатньо високу якість товарів та задоволеність клієнтів у цьому регіоні.  

Рекомендація: варто додатково проаналізувати, товари з яким ціновим діапазоном найчастіше повертаються, щоб виявити ризикові сегменти (наприклад, низький, середній або високий ціновий рівень).  


---
## Завдання 6 (Середнє). Сортування та ранжування

1. Відсортуйте датафрейм за колонкою `total` у спадному порядку та виведіть топ-10 найбільших замовлень.
2. Додайте нову колонку `rank`, що містить порядковий номер замовлення за розміром суми (від найбільшого до найменшого) за допомогою `.rank()`.

In [8]:
# Ваш код тут
top_10 = df.sort_values(by='total', ascending=False).head(10)
print(top_10)

df['rank'] = df['total'].rank(ascending=False, method='dense')


     order_id       date     category   product    city  quantity    price  \
190      1191 2023-07-10         Одяг  Товар_22    Київ         9  4924.96   
58       1059 2023-02-28     Продукти  Товар_33  Харків         9  4848.88   
186      1187 2023-07-06        Спорт  Товар_34    Київ         9  4675.12   
55       1056 2023-02-25  Електроніка   Товар_3  Дніпро         8  4912.77   
169      1170 2023-06-19     Продукти  Товар_50    Київ         8  4739.05   
191      1192 2023-07-11        Спорт  Товар_25  Харків         9  4202.55   
193      1194 2023-07-13         Одяг  Товар_22     NaN         8  4608.17   
51       1052 2023-02-21     Продукти  Товар_35  Дніпро         8  4488.99   
25       1026 2023-01-26        Книги  Товар_18  Дніпро         8  4458.11   
143      1144 2023-05-24     Продукти  Товар_31    Київ         9  3732.53   

    payment_method  rating  is_returned     total  
190        Готівка     4.0         True  44324.64  
58          Картка     5.0        Fal

**Висновок:** *(Яка максимальна сума замовлення? З яких категорій найбільші замовлення?)*  
Максимальна сума замовлення становить 44 324.64, але це замовлення було повернене, отже максимальна сума замовлення становить - 43 639.92.  

Топ-10 найбільших замовлень формуються переважно з категорій «Продукти», «Спорт», «Одяг» та «Електроніка», тобто великі чеки розподілені між кількома сегментами, а не концентруються в одній категорії.

---
## Завдання 7 (Середнє). Групування та агрегація

За допомогою `groupby()` розрахуйте для кожної категорії товарів:
- загальний обсяг продажів (сума `total`),
- кількість замовлень,
- середню ціну (`price`).

Відсортуйте результат за загальним обсягом продажів.

In [9]:
# Ваш код тут
result = df.groupby('category').agg(
    total_sales=('total', 'sum'),
    orders_count=('order_id', 'count'),
    avg_price=('price', 'mean')
).sort_values(by='total_sales', ascending=False)

print(result)

             total_sales  orders_count    avg_price
category                                           
Продукти       583761.68            49  2513.979787
Одяг           474135.61            35  2707.906176
Електроніка    428129.79            43  2334.470698
Спорт          400802.56            36  2494.205143
Книги          380844.39            37  2491.881944


**Висновок:** *(Яка категорія приносить найбільший дохід? Чи збігається це з кількістю замовлень?)*  

Найбільший дохід приносить категорія «Продукти» — 583 761.68, що робить її ключовим драйвером виручки в датасеті. Далі йдуть «Одяг» (474 135.61) та «Електроніка» (428 129.79).  

При цьому найбільша кількість замовлень не повністю збігається з лідером по доходу:  

- «Продукти» мають і найбільший дохід, і найбільшу кількість замовлень (49) → це головний сегмент,  
- «Електроніка» має багато замовлень (43), але менший дохід,
- «Одяг» має менше замовлень (35), але високий середній чек, тому займає друге місце по доходу.  

**Рекомендації**:  
- Сфокусуватися на категорії «Продукти», оскільки вона приносить найбільший дохід і має найбільшу кількість замовлень — це стабільний основний драйвер виручки.  
- Посилити монетизацію «Електроніки», адже при великій кількості замовлень її дохід нижчий — варто перевірити середній чек, структуру товарів і можливості апселу (дорожчі комплекти, допродажі).  
- Підтримувати «Одяг» як високочековий сегмент, оскільки він має менше замовлень, але високий дохід — потенціал росту тут у збільшенні кількості продажів.  
- Аналізувати цінову стратегію «Спорту» і «Книг», щоб зрозуміти, як збільшити їхній вклад у загальну виручку (акції, бандли, знижки).



---
## Завдання 8 (Середнє). Обробка пропущених значень

1. Заповніть пропущені значення в колонці `price` медіаною цін відповідної категорії (використайте `groupby().transform('median')`).
2. Заповніть пропущені значення в колонці `city` рядком `'Невідомо'`.
3. Пропущені значення в `rating` залиште без змін — поясніть чому у висновку.

Перевірте, що пропусків у `price` та `city` більше немає.

In [10]:
# Ваш код тут
df['price'] = df['price'].fillna(
    df.groupby('category')['price'].transform('median')
)

df['city'] = df['city'].fillna('Невідомо')
print(df[['price', 'city', 'rating']].isnull().sum())

price      0
city       0
rating    23
dtype: int64


**Висновок:** *(Чому медіана краща за середнє для заповнення цін? Чому пропуски в rating можна залишити?)*  

Медіана менш чутлива до дуже високих або дуже низьких значень. У наборі даних є дорогі замовлення, які можуть суттєво зміщувати середнє значення, тому медіана краще відображає типову ціну товару в категорії.  

Пропуски в колонці `rating` доцільно залишити без змін, оскільки відсутність рейтингу не означає певну оцінку. Найчастіше це свідчить про те, що клієнт не залишив відгук. Заповнення таких пропусків середнім або іншим значенням могло б спотворити результати аналізу задоволеності клієнтів і призвести до хибних висновків.

---
## Завдання 9 (Середнє). Додавання обчислюваних колонок

Додайте до датафрейму нові колонки:
1. `month` — місяць замовлення (з колонки `date`).
2. `price_category` — категорія ціни: `'Низька'` (до 500), `'Середня'` (500–2000), `'Висока'` (понад 2000). Використайте `pd.cut()` або `np.select()`.
3. `discount_total` — сума зі знижкою 15%.

In [15]:
# Ваш код тут

# 1. Місяць замовлення
df['month'] = df['date'].dt.month

# 2. Категорія ціни
conditions = [
    df['price'] < 500,
    (df['price'] >= 500) & (df['price'] <= 2000),
    df['price'] > 2000
]

choices = ['Низька', 'Середня', 'Висока']

df['price_category'] = np.select(conditions, choices, default='Невідомо')

# 3. Сума зі знижкою 15%
df['discount_total'] = df['total'] * 0.85


print(df[['date', 'month', 'price', 'price_category', 'total', 'discount_total']].head())



        date  month    price price_category     total  discount_total
0 2023-01-01      1   963.38        Середня    963.38        818.8730
1 2023-01-02      1  1086.28        Середня   6517.68       5540.0280
2 2023-01-03      1  1883.84        Середня   7535.36       6405.0560
3 2023-01-04      1  2448.39         Висока   2448.39       2081.1315
4 2023-01-05      1  3110.36         Висока  21772.52      18506.6420


In [16]:

df['price_category'] = pd.cut(
    df['price'],
    bins=[0, 500, 2000, float('inf')],
    labels=['Низька', 'Середня', 'Висока']
)

print(df['price_category'].value_counts())

price_category
Висока     121
Середня     70
Низька       9
Name: count, dtype: int64


**Висновок:** *(Який розподіл цінових категорій? Як метод pd.cut спрощує категоризацію?)*  

Після категоризації всі товари були розподілені на три цінові сегменти: «Низька», «Середня» та «Висока». Це дозволяє аналізувати продажі не лише за конкретною ціною, а й за ціновими групами, що значно спрощує сегментацію клієнтів і товарів.  

Метод `pd.cut()` автоматично розподіляє числові значення по заданих інтервалах, тому не потрібно вручну перевіряти кожну ціну та присвоювати їй категорію. Завдяки цьому дані стають більш зручними для аналізу, побудови звітів і порівняння результатів між різними ціновими сегментами.

---
## Завдання 10 (Середнє). Зведена таблиця (pivot table)

Побудуйте зведену таблицю за допомогою `pd.pivot_table()`, де:
- рядки — `city`,
- колонки — `category`,
- значення — сума `total`,
- агрегація — `sum`.

Додайте підсумки по рядках і колонках (`margins=True`).

In [17]:
# Ваш код тут
pivot = pd.pivot_table(
    df,
    values='total',
    index='city',
    columns='category',
    aggfunc='sum',
    margins=True
)

print(pivot)

category  Електроніка      Книги       Одяг   Продукти      Спорт         All
city                                                                         
Дніпро       65438.16  116118.97   31179.87  147336.56   72244.38   432317.94
Київ         58713.88   87339.03  103754.43  118283.10  113742.95   481833.39
Львів       154285.60  108959.19   74340.69   55768.80   43155.08   436509.36
Невідомо     21012.44   25984.45   46962.31   40508.03    5917.59   140384.82
Одеса        84611.65   26773.57   82361.08   60466.75   65539.34   319752.39
Харків       44068.06   15669.18  135537.23  161398.44  100203.22   456876.13
All         428129.79  380844.39  474135.61  583761.68  400802.56  2267674.03


 **Висновок:** *(Яке місто і яка категорія дають найбільший сумарний дохід? Що показують margins?)*  

 Найбільший сумарний дохід серед міст приносить Київ (≈481 833.39), що робить його ключовим регіоном за обсягом виручки. Серед товарних категорій найбільший внесок у загальний дохід має «Продукти» (≈583 761.68), що робить цю категорію головним драйвером продажів.  

З аналізу перетину видно, що окремі міста та категорії мають різну силу: наприклад, Львів сильно представлений в «Електроніці», а Харків — у «Продуктах», що свідчить про регіональні відмінності в попиті.  

`margins` дозволяє швидко побачити загальну картину бізнесу без додаткових обчислень:  

- які міста найбільш прибуткові,  
- які категорії формують основний дохід,  
- загальний оборот компанії.  


---
## Завдання 11 (Середнє). Підрахунок унікальних значень і частот

1. Підрахуйте кількість унікальних товарів (`product`) у кожній категорії.
2. Виведіть топ-5 найпопулярніших товарів за кількістю замовлень за допомогою `value_counts()`.
3. Визначте частку кожного способу оплати у відсотках (`normalize=True`).

In [18]:
# Ваш код тут
# 1. Унікальні товари у кожній категорії
unique_products = df.groupby('category')['product'].nunique()
print("Унікальні товари по категоріях:")
print(unique_products)

# 2. Топ-5 найпопулярніших товарів
top5 = df['product'].value_counts().head(5)
print("\nТоп-5 товарів:")
print(top5)

# 3. Частка способів оплати у відсотках
payment_share = df['payment_method'].value_counts(normalize=True) * 100
print("\nСпособи оплати (%):")
print(payment_share.round(2))

Унікальні товари по категоріях:
category
Електроніка    32
Книги          25
Одяг           27
Продукти       30
Спорт          25
Name: product, dtype: int64

Топ-5 товарів:
product
Товар_33    11
Товар_48     9
Товар_35     9
Товар_22     7
Товар_20     6
Name: count, dtype: int64

Способи оплати (%):
payment_method
Готівка    34.0
Онлайн     33.5
Картка     32.5
Name: proportion, dtype: float64


**Висновок:** *(Який спосіб оплати найпопулярніший? Чи рівномірний розподіл між способами?)*  

Найпопулярнішим способом оплати є готівка — її використовують 34,0% покупців. Частка оплат онлайн становить 33,5%, а карткою — 32,5%.  

Розподіл між способами оплати є майже рівномірним, оскільки різниця між найпопулярнішим і найменш популярним способом становить лише 1,5%. Це свідчить про те, що покупці приблизно однаково часто використовують усі три способи оплати, без явного домінування одного з них.  

Якщо проаналізувати способи оплати в розрізі категорій товарів, можна отримати більш глибокі висновки, наприклад:  
- переваги покупців залежно від категорії товару,  
- можливості для маркетингу,  
- оцінити популярність онлайн-продажів,  
- виявити нові канали продажу.  

---
## Завдання 12 (Середнє). Аналіз повернень

Розрахуйте рівень повернень (частку повернених замовлень) для кожної категорії та кожного міста. Визначте:
- категорію з найвищим рівнем повернень,
- місто з найнижчим рівнем повернень.

Підказка: використайте `groupby()['is_returned'].mean()`.

In [19]:
# Ваш код тут
# Рівень повернень по категоріях
category_returns = df.groupby('category')['is_returned'].mean() * 100

# Рівень повернень по містах
city_returns = df.groupby('city')['is_returned'].mean() * 100

# Категорія з найвищим рівнем
top_category = category_returns.idxmax()

# Місто з найнижчим рівнем
low_city = city_returns.idxmin()

print("Рівень повернень по категоріях (%):")
print(category_returns.round(2).sort_values(ascending=False))

print(f"\nКатегорія з найвищим рівнем повернень: {top_category} ({category_returns[top_category]:.2f}%)")
print(f"Місто з найнижчим рівнем повернень:    {low_city} ({city_returns[low_city]:.2f}%)")

Рівень повернень по категоріях (%):
category
Одяг           17.14
Електроніка    16.28
Книги          13.51
Спорт          11.11
Продукти        6.12
Name: is_returned, dtype: float64

Категорія з найвищим рівнем повернень: Одяг (17.14%)
Місто з найнижчим рівнем повернень:    Дніпро (5.26%)


**Висновок:** *(Яка категорія найчастіше повертається? Про що це може свідчити?)*  
Найвищий рівень повернень спостерігається в категорії «Одяг» — 17,14% від усіх продажів цієї категорії. Друге місце займає «Електроніка» з показником 16,28%. Найнижчий рівень повернень мають «Продукти» — лише 6,12%.  

Високий рівень повернень одягу може свідчити про те, що покупцям складно правильно підібрати розмір, фасон, колір або модель під час замовлення. Для електроніки причинами повернень можуть бути невідповідність очікуванням, складність використання або технічні несправності.  

Низький рівень повернень продуктів харчування може пояснюватися тим, що такі товари купуються регулярно, мають зрозумілі характеристики та часто не підлягають поверненню після покупки.  

Також варто зазначити, що Дніпро має найнижчий рівень повернень (5,26%), що може свідчити про вищу задоволеність клієнтів або особливості купівельної поведінки в цьому місті. Загалом аналіз повернень допомагає виявити категорії товарів, які потребують покращення описів, якості або процесу підбору для покупців.

---
## Завдання 13 (Підвищена складність). Часовий аналіз

1. Встановіть колонку `date` як індекс датафрейму.
2. Розрахуйте щомісячний обсяг продажів (ресемплінг `resample('ME')` або `resample('M')`, sum по `total`).
3. Знайдіть найкращий і найгірший місяці за продажами.
4. Поверніть звичайний числовий індекс (`reset_index()`).

In [21]:
# Ваш код тут
# 1.
df = df.set_index('date')

# 2.
monthly_sales = df['total'].resample('ME').sum()

# 3.
best_month = monthly_sales.idxmax()
worst_month = monthly_sales.idxmin()

print("Щомісячні продажі:")
print(monthly_sales)
print(f"\nНайкращий місяць: {best_month.strftime('%Y-%m')} — {monthly_sales[best_month]:,.2f}")
print(f"Найгірший місяць: {worst_month.strftime('%Y-%m')} — {monthly_sales[worst_month]:,.2f}")

# 4.
df = df.reset_index()

Щомісячні продажі:
date
2023-01-31    326994.68
2023-02-28    341104.20
2023-03-31    292529.84
2023-04-30    305085.53
2023-05-31    335619.62
2023-06-30    386464.61
2023-07-31    279875.55
Freq: ME, Name: total, dtype: float64

Найкращий місяць: 2023-06 — 386,464.61
Найгірший місяць: 2023-07 — 279,875.55


**Висновок:** *(Який місяць виявився найуспішнішим? Чи є сезонність у даних?)*  

Найуспішнішим місяцем за обсягом продажів став червень 2023 року, коли загальна сума продажів склала 386 464,61 грн. Найнижчий показник зафіксовано у липні 2023 року — 279 875,55 грн.  

За даними спостерігаються певні коливання обсягів продажів протягом року: після відносно стабільних показників у першій половині року продажі зростають і досягають максимуму в червні, а вже в липні помітно знижуються. Це може свідчити про вплив сезонних факторів, маркетингових активностей або зміну попиту в літній період.  

Однак для впевненого висновку щодо сезонності даних недостатньо одного неповного року спостережень (лише 7 місяців). Для виявлення сезонних закономірностей бажано проаналізувати дані щонайменше за кілька років. На основі поточних даних можна лише припустити наявність короткострокових коливань попиту, але не підтвердити стійку сезонність.  


---
## Завдання 14 (Підвищена складність). Накопичувальні показники

1. Відсортуйте датафрейм за датою.
2. Додайте колонку `cumulative_total` — накопичувальний обсяг продажів з початку року (`cumsum()`).
3. Визначте, в який день було досягнуто 50% від загального річного обсягу продажів.

In [23]:
# Ваш код тут
# 1. Сортуємо датафрейм за датою
df = df.sort_values('date').reset_index(drop=True)

# 2. Накопичувальний обсяг продажів з початку року
df['cumulative_total'] = df['total'].cumsum()

# 3. День досягнення 50% річного обсягу продажів
annual_total = df['total'].sum()
halfway = annual_total * 0.5

milestone_row = df[df['cumulative_total'] >= halfway].iloc[0]
milestone_date = milestone_row['date']

print(f"Загальний річний обсяг: {annual_total:,.2f}")
print(f"50% порогове значення:  {halfway:,.2f}")
print(f"Дата досягнення 50%:    {milestone_date.strftime('%Y-%m-%d')}")
print(f"Накопичено на цей день: {milestone_row['cumulative_total']:,.2f}")

print("\nПерші рядки з cumulative_total:")
print(df[['date', 'total', 'cumulative_total']].head())

Загальний річний обсяг: 2,267,674.03
50% порогове значення:  1,133,837.01
Дата досягнення 50%:    2023-04-15
Накопичено на цей день: 1,141,918.50

Перші рядки з cumulative_total:
        date     total  cumulative_total
0 2023-01-01    963.38            963.38
1 2023-01-02   6517.68           7481.06
2 2023-01-03   7535.36          15016.42
3 2023-01-04   2448.39          17464.81
4 2023-01-05  21772.52          39237.33


**Висновок:** *(Коли приблизно була зароблена половина річного доходу? Що це означає для бізнесу?)*  

Половина річного доходу (1 133 837,01 грн) була досягнута 15 квітня 2023 року. На цю дату накопичений дохід склав 1 141 918,50 грн.  

Це свідчить про високі темпи продажів на початку року. Оскільки 50% доходу було отримано менш ніж за чотири місяці, можна припустити, що в подальші місяці темпи продажів дещо знизилися. Для бізнесу це сигнал про необхідність аналізу факторів, які вплинули на таку динаміку. Це може бути як нормою (наприклад: сезонність), так і відхиленням від норми.  

---
## Завдання 15 (Підвищена складність). apply() та lambda

1. Використайте `apply()` з lambda-функцією для створення колонки `total_with_tax`, де ПДВ становить 20%, але лише для замовлень на суму понад 1000 (решта — без змін).
2. Напишіть функцію `classify_rating(r)`, що повертає `'Поганий'` (1–2), `'Нейтральний'` (3), `'Хороший'` (4–5), `'Без оцінки'` (NaN). Застосуйте її до колонки `rating`.

In [24]:
# Ваш код тут
# 1. total_with_tax через lambda
df['total_with_tax'] = df['total'].apply(
    lambda x: x * 1.2 if x > 1000 else x
)

# 2. Функція класифікації рейтингу
def classify_rating(r):
    if pd.isna(r):
        return 'Без оцінки'
    elif r <= 2:
        return 'Поганий'
    elif r == 3:
        return 'Нейтральний'
    else:
        return 'Хороший'

df['rating_category'] = df['rating'].apply(classify_rating)

print(df[['total', 'total_with_tax', 'rating', 'rating_category']].head(10))
print("\nРозподіл рейтингів:")
print(df['rating_category'].value_counts())

      total  total_with_tax  rating rating_category
0    963.38         963.380     5.0         Хороший
1   6517.68        7821.216     4.0         Хороший
2   7535.36        9042.432     5.0         Хороший
3   2448.39        2938.068     5.0         Хороший
4  21772.52       26127.024     3.0     Нейтральний
5  16885.08       20262.096     NaN      Без оцінки
6   9358.20       11229.840     4.0         Хороший
7  14999.92       17999.904     4.0         Хороший
8   1389.48        1667.376     4.0         Хороший
9   3898.68        4678.416     1.0         Поганий

Розподіл рейтингів:
rating_category
Хороший        116
Нейтральний     35
Поганий         26
Без оцінки      23
Name: count, dtype: int64


**Висновок:** *(Яка частка замовлень має хороший рейтинг? Коли apply кращий за vectorized-операції?)*  

Більшість замовлень мають хороший рейтинг — 116 із 200 замовлень, або 58% від загальної кількості. Частка замовлень із нейтральним рейтингом становить 17,5%, з поганим — 13%, а 11,5% замовлень не мають оцінки.  

Це свідчить про те, що переважна частина клієнтів задоволена покупками, хоча є певна кількість негативних відгуків, які варто додатково проаналізувати.  

*Коли `apply()` кращий за vectorized-операції?*  

`apply()` кращий коли потрібно виконати складну логіку або кілька умов для кожного рядка, або коли необхідно перетворювати значення за допомогою власної функції, також якщо немає готової векторизованої операції для потрібного завдання.


---
## Завдання 16 (Підвищена складність). Злиття датафреймів (merge)

Створіть додатковий датафрейм `df_discounts` з колонками `category` та `discount_pct` (різні знижки для кожної категорії). Об'єднайте його з основним датафреймом за допомогою `pd.merge()` та розрахуйте `final_total` — суму після застосування відповідної знижки.

```python
df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})
```

In [25]:
df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})

# Ваш код тут

# Merge основного датафрейму з таблицею знижок
df = pd.merge(df, df_discounts, on='category', how='left')

# Розраховуємо фінальну суму після знижки
df['final_total'] = df['total'] * (1 - df['discount_pct'].fillna(0))

print(df[['category', 'total', 'discount_pct', 'final_total']].head(10))
print("\nПропущені знижки (категорії не з таблиці):")
print(df[df['discount_pct'].isna()]['category'].value_counts())

   category     total  discount_pct  final_total
0  Продукти    963.38          0.00     963.3800
1     Спорт   6517.68          0.15    5540.0280
2     Книги   7535.36          0.05    7158.5920
3     Спорт   2448.39          0.15    2081.1315
4     Спорт  21772.52          0.15   18506.6420
5      Одяг  16885.08          0.20   13508.0640
6     Книги   9358.20          0.05    8890.2900
7     Книги  14999.92          0.05   14249.9240
8     Книги   1389.48          0.05    1320.0060
9     Спорт   3898.68          0.15    3313.8780

Пропущені знижки (категорії не з таблиці):
Series([], Name: count, dtype: int64)


**Висновок:** *(Яка категорія отримала найбільшу сумарну знижку? Коли варто використовувати merge замість map?)*  

Найбільшу сумарну знижку, ймовірно, отримали категорії «Одяг» (20%) і «Спорт» (15%), бо в них найвищі ставки знижок і великі суми замовлень.  

`merge()` варто використовувати, коли потрібно:  

- додати кілька колонок з іншої таблиці,  
- об’єднати два DataFrame.  

`map()` — коли потрібно швидко додати одну колонку за відповідністю ключ → значення.

---
## Завдання 17 (Підвищена складність). Групові трансформації

1. Додайте колонку `city_avg_total` — середня сума замовлення по місту (для кожного рядка підставте середнє по його місту), використайте `groupby().transform('mean')`.
2. Додайте колонку `above_city_avg` — булева колонка, що показує, чи перевищує замовлення середнє по місту.
3. Підрахуйте частку таких замовлень для кожного міста.

In [26]:
# Ваш код тут
# 1. Середня сума по місту для кожного рядка
df['city_avg_total'] = df.groupby('city')['total'].transform('mean')

# 2. Чи перевищує замовлення середнє по місту
df['above_city_avg'] = df['total'] > df['city_avg_total']

# 3. Частка замовлень вище середнього по кожному місту
city_above_avg = df.groupby('city')['above_city_avg'].mean() * 100

print(df[['city', 'total', 'city_avg_total', 'above_city_avg']].head(10))
print("\nЧастка замовлень вище середнього по містах (%):")
print(city_above_avg.round(2).sort_values(ascending=False))

     city     total  city_avg_total  above_city_avg
0    Київ    963.38    13022.524054           False
1   Львів   6517.68     9700.208000           False
2  Дніпро   7535.36    11376.787895           False
3  Дніпро   2448.39    11376.787895           False
4  Дніпро  21772.52    11376.787895            True
5   Львів  16885.08     9700.208000            True
6  Дніпро   9358.20    11376.787895           False
7   Львів  14999.92     9700.208000            True
8  Дніпро   1389.48    11376.787895           False
9   Одеса   3898.68    11025.944483           False

Частка замовлень вище середнього по містах (%):
city
Дніпро      42.11
Невідомо    40.00
Одеса       40.00
Харків      38.46
Київ        35.14
Львів       32.61
Name: above_city_avg, dtype: float64


**Висновок:** *(Чим transform відрізняється від звичайного groupby? У якому місті найбільша частка великих замовлень?)*  

Найбільша частка замовлень вище середнього по місту у Дніпрі — 42,11%. Це означає, що майже половина покупок у цьому місті є відносно великими.  

`transform()` відрізняється від `groupby()` тим, що:  

`groupby()` повертає агреговані значення (по групах),  
`transform()` повертає результат для кожного рядка, зберігаючи початковий розмір DataFrame (тобто “приклеює” групову статистику назад до кожного запису).

---
## Завдання 18 (Складне). Ковзне середнє та аномалії

1. Згрупуйте дані за місяцем та розрахуйте щоденну виручку.
2. Розрахуйте 7-денне ковзне середнє виручки (`rolling(7).mean()`).
3. Визначте дні-аномалії: дні, коли виручка відхилилася від ковзного середнього більш ніж на 2 стандартних відхилення.

In [27]:
# Ваш код тут
# 1. Щоденна виручка
daily_revenue = df.groupby('date')['total'].sum().reset_index()
daily_revenue.columns = ['date', 'revenue']
daily_revenue = daily_revenue.sort_values('date')

# 2. 7-денне ковзне середнє
daily_revenue['rolling_mean'] = daily_revenue['revenue'].rolling(7).mean()
daily_revenue['rolling_std']  = daily_revenue['revenue'].rolling(7).std()

# 3. Аномалії — відхилення більш ніж на 2 std
daily_revenue['is_anomaly'] = (
    (daily_revenue['revenue'] - daily_revenue['rolling_mean']).abs()
    > 2 * daily_revenue['rolling_std']
)

anomalies = daily_revenue[daily_revenue['is_anomaly']]

print(f"Знайдено аномалій: {len(anomalies)}")
print("\nДні-аномалії:")
print(anomalies[['date', 'revenue', 'rolling_mean', 'rolling_std']].round(2))

Знайдено аномалій: 4

Дні-аномалії:
          date   revenue  rolling_mean  rolling_std
25  2023-01-26  35664.88       8803.79     12716.33
51  2023-02-21  35911.92      10910.35     11346.27
85  2023-03-27  29701.80       8022.48     10634.73
143 2023-05-24  33592.77       9505.60     10875.75


**Висновок:** *(Скільки аномальних днів виявлено? Що може бути причиною різких відхилень виручки?)*  

Виявлено 4 аномальні дні з різкими відхиленнями виручки.  

Такі сплески можуть бути спричинені:  

- акціями або розпродажами,  
- великими разовими замовленнями (B2B-клієнти),  
- святковими періодами,  
- або можливими статистичними викидами/помилками в даних.  

Для точного розуміння причин варто зробити аналіз маркетингових активностей та подій у бізнесі.

---
## Завдання 19 (Складне). Когортний аналіз

Виконайте спрощений когортний аналіз за місяцем першого замовлення:
1. Для кожного товару (`product`) знайдіть місяць першого продажу — це його «когорта».
2. Об'єднайте цю інформацію з основним датафреймом.
3. Побудуйте зведену таблицю: когорта (місяць першого продажу) проти місяця замовлення, значення — кількість замовлень.

Підказка: використайте `groupby('product')['date'].transform('min')` та `dt.to_period('M')`.

In [28]:
# Ваш код тут
# 1. Місяць першого продажу для кожного товару (когорта)
df['cohort'] = (
    df.groupby('product')['date']
    .transform('min')
    .dt.to_period('M')
)

# 2. Місяць поточного замовлення
df['order_month'] = df['date'].dt.to_period('M')

# 3. Зведена таблиця: когорта × місяць замовлення
cohort_table = df.pivot_table(
    index='cohort',
    columns='order_month',
    values='product',
    aggfunc='count'
).fillna(0).astype(int)

print("Коґортна таблиця (кількість замовлень):")
print(cohort_table)

print(f"\nРозмір таблиці: {cohort_table.shape}")
print(f"Когорт: {cohort_table.shape[0]}, Місяців: {cohort_table.shape[1]}")

Коґортна таблиця (кількість замовлень):
order_month  2023-01  2023-02  2023-03  2023-04  2023-05  2023-06  2023-07
cohort                                                                    
2023-01           31       15       18       11       14       15        8
2023-02            0       13        6        6       12        8        7
2023-03            0        0        7        6        4        2        1
2023-04            0        0        0        7        1        3        0
2023-06            0        0        0        0        0        2        1
2023-07            0        0        0        0        0        0        2

Розмір таблиці: (6, 7)
Когорт: 6, Місяців: 7


**Висновок:** *(Що показує когортна таблиця? Чи продовжують товари з першої когорти продаватися в наступних місяцях?)*  

'Когортна' таблиця показує, як клієнти, які вперше купили в певному місяці, роблять покупки в наступні місяці.  

Клієнти з когорти 2023-01 далі ще купували в інших місяцях, тобто вони повертаються. Але з кожною новою когортою кількість повторних покупок менша.  

Отже, товари з першої 'когорти' продовжують купуватися в наступні місяці, але активність клієнтів поступово знижується.  

---
## Завдання 20 (Складне). Підсумковий звіт

Сформуйте підсумковий звіт у вигляді одного датафрейму `summary_report`, що містить для кожної категорії:
- загальний дохід (`total_revenue`),
- кількість замовлень (`order_count`),
- середній чек (`avg_order_value`),
- частку доходу від загального (`revenue_share_pct`),
- рівень повернень у відсотках (`return_rate_pct`),
- середній рейтинг без NaN (`avg_rating`).

Відсортуйте за `total_revenue` спадно. Збережіть результат у файл `summary_report.csv`.

In [29]:
# Ваш код тут
summary_report = (
    df.groupby('category')
    .agg(
        total_revenue   = ('total',       'sum'),
        order_count     = ('total',       'count'),
        avg_order_value = ('total',       'mean'),
        return_rate_pct = ('is_returned', 'mean'),
        avg_rating      = ('rating',      'mean'),
    )
    .reset_index()
)

# Частка доходу від загального
summary_report['revenue_share_pct'] = (
    summary_report['total_revenue'] / summary_report['total_revenue'].sum() * 100
)

# Повернення у відсотки та округлення
summary_report['return_rate_pct'] = summary_report['return_rate_pct'] * 100

summary_report = (
    summary_report
    .sort_values('total_revenue', ascending=False)
    .reset_index(drop=True)
    .round(2)
)

print(summary_report.to_string(index=False))

# Збереження у CSV
summary_report.to_csv('summary_report.csv', index=False, encoding='utf-8-sig')
print("\nФайл збережено: summary_report.csv")

   category  total_revenue  order_count  avg_order_value  return_rate_pct  avg_rating  revenue_share_pct
   Продукти      583761.68           47         12420.46             6.12        3.91              25.74
       Одяг      474135.61           34         13945.16            17.14        3.97              20.91
Електроніка      428129.79           43          9956.51            16.28        3.64              18.88
      Спорт      400802.56           35         11451.50            11.11        3.52              17.67
      Книги      380844.39           36         10579.01            13.51        4.03              16.79

Файл збережено: summary_report.csv


**Висновок:** *(Яка категорія є лідером за доходом? Чи є зв'язок між рейтингом та рівнем повернень? Які рекомендації можна дати бізнесу на основі цього звіту?)*  

Лідером за доходом є категорія «Продукти» — 583 761,68 грн (25,74% загального доходу).  

Зв’язок між рейтингом і поверненнями частково простежується:  
категорії з нижчим рейтингом (наприклад, «Спорт» і «Електроніка») мають вищий рівень повернень, тоді як у «Продуктів» рейтинг вищий і повернень менше.  

**Рекомендації:**

- покращити якість або опис товарів у «Електроніці» та «Спорті» (через високі повернення);  
- підтримувати сильні позиції «Продуктів» як основного джерела доходу;  
- дослідити причини низького рейтингу в окремих категоріях;  
- зменшувати повернення через кращу інформацію про товар і контроль якості.

---
## Загальні підсумки

Після виконання всіх завдань дайте відповідь на такі питання:

1. Які методи pandas виявилися для вас найбільш корисними?
2. Які труднощі виникли під час роботи з пропущеними значеннями?
3. Що можна додати до аналізу, щоб зробити його більш повним?

*(Ваші відповіді тут)*  

1. Найбільш корисні методи pandas:  
Найбільш корисними були `groupby()`, `merge()`, `transform()`, `apply()` та робота з агрегаціями `(agg())`. Вони дозволили швидко отримувати підсумки по категоріях, містах і когортах, а також порівнювати значення між групами.  

2. Труднощі з пропущеними значеннями:  
Основна складність була у правильній обробці пропусків (наприклад, рейтинги «Без оцінки»). Важливо було не втратити дані при аналізі та правильно інтерпретувати такі значення, а не просто видаляти їх.  

3. Що можна додати до аналізу:  
Можна розширити аналіз сезонністю (за кілька років), поведінкою клієнтів (retention, LTV), а також детальнішим аналізом причин повернень і ефективності маркетингових кампаній.